File: Presentation/ClientApp/src/app/Models/Invoice.ts
import { InvoiceDetail } from './InvoiceDetail';
import { Product } from './product.model';

export class Invoice {

    idInvoice: string;
    subtotal = 0;
    totalIva = 0;
    total = 0;
    saleDate: string;
    dueDate: string;
    paymentMethod: string;
    idClient: string;
    invoiceDetails: InvoiceDetail[];

    constructor () {
        this.invoiceDetails = [];
        const fecha = new Date();
        this.saleDate = fecha.toLocaleString();
        this.dueDate =  new Date(fecha.getFullYear(), fecha.getMonth() + 2 , fecha.getDay(),
                                fecha.getHours(), fecha.getMinutes(), fecha.getSeconds()).toLocaleString();
    }

    addInvoiceDetails (product: Product, quantity: number, discount: number, price: number ) {
      const detail = new InvoiceDetail(product, quantity, discount, price);
      detail.idInvoice = this.idInvoice;
      this.invoiceDetails.push(detail);
    }

    calculateSubtotal(): void {
        this.subtotal = 0;
        for (const item of this.invoiceDetails) {
            this.subtotal += item.tolalDetail;
        }
    }

    calcularteTotalIva(): void {
        this.totalIva = 0;
        for (const item of this.invoiceDetails) {
            this.totalIva +=  item.CalculateIva();
        }
    }

    calculateTotal(): void {
        this.total = 0;
        this.calculateSubtotal();
        this.calcularteTotalIva();

        this.total = this.subtotal + this.totalIva;
    }
}

File: Entity/Invoice.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;

namespace Entity
{
    public partial class Invoice
    {
        [Key]
        [Column(TypeName= "nvarchar(4)")]
        public string IdInvoice{ get; set; }
        [Column(TypeName = "decimal(18,2)")]
        public decimal Subtotal { get; set; }
        [Column(TypeName = "decimal(18,2)")]
        public decimal TotalIva { get; set; }
        [Column(TypeName = "decimal(18,2)")]
        public decimal Total { get; set; }
        [Column(TypeName= "nvarchar(30)")]
        public string PaymentMethod { get; set; }
        [Column(TypeName= "nvarchar(30)")]
        public string SaleDate { get; set; }
        [Column(TypeName= "nvarchar(30)")]
        public string DueDate { get; set; }

        [Column(TypeName = "nvarchar(11)")]
        public string IdClient { get; set; }

        [NotMapped]
        public Client Client { get; set; }



        [Column(TypeName = "nvarchar(4)")]
        [ForeignKey("IdInvoice")]
        public virtual IList<InvoiceDetail> InvoiceDetails { get; set; } 

        public Invoice()
        {
            InvoiceDetails = new List<InvoiceDetail>();
        }

        public void AddInvoiceDetails(Product product, float quantity, float discount, decimal price)
        {
            InvoiceDetail invoiceDetail = new InvoiceDetail(product, quantity, discount, price);
            invoiceDetail.InvoiceIdInvoice = this.IdInvoice;
            InvoiceDetails.Add(invoiceDetail);
        }

        public void CalculateSubtotal()
        {
            Subtotal = InvoiceDetails.Sum(d => d.TolalDetail);
        }

        public void CalcularTotalIva()
        {
            TotalIva = InvoiceDetails.Sum(d => d.CalculateIva());
        }

        public void CalculateTotal()
        {
            CalculateSubtotal();
            CalcularTotalIva();
           
            Total = Subtotal + TotalIva ;
        }

    }
}